datalad_catalog/workflows.py

```diff
@@ -0,0 +1,248 @@
+        )
+
+    # Set super dataset of catalog
+    cat.main_id = ds.id
+    # - sync possible adjusted branch
+    ds.repo.localsync()
+    # - account for possibility of being on adjusted branch:
+    cat.main_version = ds.repo.get_hexsha(ds.repo.get_corresponding_branch())
+    cat.set_main_dataset()
+
+
+def update_workflow(superds_path, subds_path, catalog: WebCatalog):
+    """Run an update workflow on a specific subdataset and its parent
+
+    The workflow includes running several steps on the super- and subdataset:
+    - dataset- and file-level metadata extraction
+    - extracted metadata translation
+    - adding translated metadata to the catalog
+
+    It then resets the catalog's superdataset to the latest id and version of
+    the parent dataset.
+
+    This workflow assumes:
+    - The subdataset has already been added as a submodule to the parent dataset
+    - The parent dataset already contains the subdataset commit
+    """
+    # Install super and subdatasets
+    super_ds = Dataset(superds_path)
+    sub_ds = Dataset(subds_path)
+    # Dataset workflow for super+subdataset
+    yield dataset_workflow(super_ds, catalog)
+    yield dataset_workflow(sub_ds, catalog)
+    # Set super dataset of catalog
+    catalog.main_id = super_ds.id
+    # - sync possible adjusted branch
+    super_ds.repo.localsync()
+    # - account for possibility of being on adjusted branch:
+    catalog.main_version = super_ds.repo.get_hexsha(
+        super_ds.repo.get_corresponding_branch()
+    )
+    catalog.set_main_dataset()
+
+
+def dataset_workflow(ds: Dataset, catalog, **kwargs):
+    """Run a dataset-specific catalog generation workflow.
+
+    This includes:
+      - dataset- and file-level metadata extraction
+      - extracted metadata translation
+      - adding translated metadata to a catalog"""
+    # 1. Run dataset-level extraction
+    extracted_file = Path(ds.path) / "extracted_meta.json"
+    for name in extractor_names_dataset:
+        if check_required_files(ds, name):
+            metadata_record = extract_dataset_level(ds, name)
+            write_jsonline_to_file(extracted_file, metadata_record)
+    # 2. Run file-level extraction, add output to same file
+    # - Not implemented yet
+    # 3. Run translation
+    translated_file = Path(ds.path) / "translated_meta.json"
+    with open(extracted_file) as file:
+        for line in file:
+            meta_dict = json.loads(line.rstrip())
+            try:
+                extr_name = meta_dict["extractor_name"]
+                extr_type = meta_dict["type"]
+                mapping_path = schema_dir / get_translation_map(
+                    extr_name, extr_type
+                )
+                write_jsonline_to_file(
+                    translated_file,
+                    translate_to_catalog(meta_dict, mapping_path),
+                )
+            except Exception as e:
+                raise (e)
```

Comment:
If I understand correctly, the try & catch currently changes nothing (because we re-raise immediately). Was it meant to  provide a placeholder for handling in the future?

Reply:
That's what I typically do to account for future exception handling yes, but I do not recall what I had in mind for the future in this specific case :) Looking at it now, I think it might be a good idea to have the exception (or rather warning?) part of the result record, but to continue on to the next line (i.e. `pass`) in case the `try` fails. I am not immediately sure how to do both at the same time.

Reply:
Went with:
```python
except Exception as e:
    lgr.error("Failed to translate line due to error: %s", str(e))
    continue
```